named_arrays/_functions/functions.py

```diff
@@ -717,3 +721,196 @@ def __setitem__(
+            start=na.SpectralPositionalVectorArray(
+                wavelength=610 * u.AA,
+                position=na.Cartesian2dVectorArray(
+                    x=-50 * u.arcsec,
+                    y=-50 * u.arcsec,
+                )
+            ),
+            stop=na.SpectralPositionalVectorArray(
+                wavelength=650 * u.AA,
+                position=na.Cartesian2dVectorArray(
+                    x=50 * u.arcsec,
+                    y=50 * u.arcsec,
+                )
+            ),
+            num=na.SpectralPositionalVectorArray(
+                wavelength=5,
+                position=na.Cartesian2dVectorArray(
+                    x=51,
+                    y=51,
+                ),
+            ),
+            axis=na.SpectralPositionalVectorArray(
+                wavelength='wavelength',
+                position=na.Cartesian2dVectorArray(
+                    x='x',
+                    y='y',
+                ),
+            ),
+        )
+            
+        outputs = na.Cartesian2dVectorArray(
+            x=1 * u.mm + (u.mm / u.arcsec) * inputs.position.x + .001 * u.mm / (u.arcsec ** 2) * (
+                inputs.position.y) ** 2 - 1 * (
+                      u.mm / u.AA) * inputs.wavelength,
+            y=5 * u.mm + na.ScalarLinearSpace(-10 * u.s, 10 * u.s, num=3, axis='time') * (
+                    u.mm / (u.arcsec * u.s)) * inputs.position.y + .05 * u.mm / (u.arcsec ** 2) * (inputs.position.x) ** 2,
+        )
+            
+    These outputs are a second degree polynomial of position_x and postion_y, and have coefficients that depend on wavelength
+    and time.
+    
+    We can fit these inputs and outputs with a second degree polynomial in position_x and position_y to demonstrate we 
+    get the same answer back.
+    
+    .. jupyter-execute::
+
+        fit = na.PolynomialFitFunctionArray(
+            inputs=inputs,
+            outputs=outputs,
+            degree=2,
+            components_polynomial=('position_x', 'position_y'),
+            axis_polynomial=('x', 'y'),
+        )
+
+        fit.coefficients
+
+    If one instead fits a linear model, the error between the fit, and the data can be easily visualized
+    
+    """
+
+    @functools.cached_property
+    def coefficients(self) -> na.AbstractVectorArray | na.AbstractMatrixArray:
+        d = self.design_matrix(self.inputs)
+        dTd = self._outer(d, d, self.axis_polynomial)
+        dTo = self._outer(d, self.outputs, self.axis_polynomial)
+
+        return dTd.inverse @ dTo
+
+    def design_matrix(self, inputs):
+        design_matrix = {}
+        cartesian_nd = inputs.cartesian_nd.broadcasted
+        # grab subset of components involved in polynomial fit
+        cartesian_nd = na.CartesianNdVectorArray(
+            {
+                k: cartesian_nd.components[k]
+                for k in cartesian_nd.components.keys() & self.components_polynomial
+            }
+        )
```

Comment:
Since all you do with this is just access the components later, maybe just leave as a dict?